### PR TITLE
KAFKA-3708: rethink exception handling in Kafka Streams

### PR DESCRIPTION
**Pull request by dguy**
KafkaExceptions currently thrown from within StreamThread/StreamTask currently bubble up without any additional context. This makes it hard to figure out where something went wrong, i.e, which topic had the serialization exception etc

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang - i've been looking at the exception handling. Beyond catching and adding some more context to the exceptions i'm not sure what else we can do.
This is by no means complete, but wanted to get your feedback before wasting too much time on this.

**Comment by guozhangwang** [comment]
@dguy The only exception I was thinking about for post-KIP4 is `consumer.commitSync`, when it throws `CommitFailedException` which usually indicate a rebalance happens behind the scene and this task is no longer owned by the consumer. Today we throw this exception all the way to the users as well but we can actually auto-handle it by just closing the task since we know it has migrated to another instance already.

We can probably do this in a different JIRA, so I will just move ahead to review the current PR as you are working to complete this WIP.

**Comment by dguy** [comment]
@guozhangwang - looks like we just log on `CommitFailedException`

```
try {
            task.commit();
        } catch (CommitFailedException e) {
            // commit failed. Just log it.
            log.warn("stream-thread [{}] Failed to commit {} #{}", this.getName(), task.getClass().getSimpleName(), task.id(), e);
        }
```

**Comment by guozhangwang** [comment]
There are a few places that we should handle exceptions better:
1. In `AbstractTask.initializeOffsetLimits()`, when `consumer.committed()` throws exceptions except `WakeupException` we could wrap them with `ProcessorStateException` indicating that something went wrong when initializing the offset limit.
2. In `ProcessorContextImpl / StandbyContextImpl` constructor, if `config.keySerde() / valueSerde()` throws any exceptions, wrap them with `StreamsException` indicating something went wrong when configuring the default serializers or deserializers.
3. In `RecordCollector`, there are calls to `producer.send()` throws `TimeoutException`, we should consider backing off and retry.
4. In `ProcessorStateManager.register()`, when calling `consumer.partitionsFor` throwing `TimeoutException`, we could wrap it with the same `StreamsException(String.format("task [%s]  Could not find partition info for topic: %s", taskId, topic));`.
5. In `ProcessorStateManager.updateStandbyStates()`, when calling `restoreCallback.restore()` we should capture any thrown exceptions and wrap them with a different messages indicating that something went wrong while restoring the state store.
6. In `ProcessorStateManager.flush()`, when calling `store.flush()` capture any thrown exception and wrap with informative message; similarly for `store.flush()` and `store.close()` in the `close()` function.
7. In `ProcessorNode`, when calling underlying `processor.init()` and `processor.process()` and `processor.close`, we could wrap any thrown exceptions indicating which processor node of which tasks went wrong, etc.
8. In `StreamTask.punctuate()`, capture any thrown exceptions indicating which thread's which tasks went wrong, etc.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 1]
```diff
@@ -234,6 +245,8 @@ public void punctuate(ProcessorNode node, long timestamp) {
         updateProcessorContext(createRecordContext(stampedRecord), node);
         try {
             node.processor().punctuate(timestamp);
+        } catch (KafkaException ke) {
+            throw new StreamsException("exception caught in punctuate. taskId=" + id + " processor=" + node.name(), ke);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Use String.format as well?

**Comment by guozhangwang** [comment]
@dguy I wrote the above meta comment before reviewing the diff, and I think some of them may be already addressed.

**Comment by dguy** [comment]
thanks @guozhangwang - i've been through your list above and added more informative exceptions.  w.r.t. backoff and retry in `RecordCollector` - how long do we backoff for and how many attempts? I've set it to 100ms backoff and 3 attempts. But, do we want to make this configurable?

**Comment by guozhangwang** [comment]
@dguy I think it is good enough for now and we can consider making it configurable in the future. There are other places such as `ProcesssStateManager.restore` where we also use hard-written backoffs for metadata refresh.

**Comment by guozhangwang** [comment]
BTW could you rebase?

**Comment by dguy** [comment]
Thanks. done

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 2]
```diff
@@ -158,7 +160,12 @@ public void register(StateStore store, boolean loggingEnabled, StateRestoreCallb
                 // ignore
             }
 
-            List<PartitionInfo> partitionInfos = restoreConsumer.partitionsFor(topic);
+            List<PartitionInfo> partitionInfos = null;
+            try {
+                partitionInfos = restoreConsumer.partitionsFor(topic);
+            } catch (TimeoutException e) {
+                throw new StreamsException(String.format("task [%s]  Could not find partition info for topic: %s", taskId, topic));
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: nit: unnecessary extra space.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 3]
```diff
@@ -268,7 +275,11 @@ private void restoreActiveState(String topicName, StateRestoreCallback stateRest
         int count = 0;
         for (ConsumerRecord<byte[], byte[]> record : records) {
             if (record.offset() < limit) {
-                restoreCallback.restore(record.key(), record.value());
+                try {
+                    restoreCallback.restore(record.key(), record.value());
+                } catch (Exception e) {
+                    throw new ProcessorStateException(String.format("exception caught while trying to restore state from %s", storePartition), e);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: nit: prefix "task [%s] "

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 4]
```diff
@@ -305,7 +316,11 @@ public void flush(final InternalProcessorContext context) {
                 if (processorNode != null) {
                     context.setCurrentNode(processorNode);
                 }
-                store.flush();
+                try {
+                    store.flush();
+                } catch (Exception e) {
+                    throw new ProcessorStateException(String.format("Failed to flush state store %s", store.name()), e);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: Ditto here.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 5]
```diff
@@ -321,8 +336,12 @@ public void close(Map<TopicPartition, Long> ackedOffsets) throws IOException {
                 log.debug("task [{}] Closing stores.", taskId);
                 for (Map.Entry<String, StateStore> entry : stores.entrySet()) {
                     log.debug("task [{}} Closing storage engine {}", taskId, entry.getKey());
-                    entry.getValue().flush();
-                    entry.getValue().close();
+                    try {
+                        entry.getValue().flush();
+                        entry.getValue().close();
+                    } catch (Exception e) {
+                        throw new ProcessorStateException(String.format("Failed to close state store %s", entry.getKey()), e);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: Ditto here.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollector.java [hunk 6]
```diff
@@ -76,18 +81,30 @@ public RecordCollector(Producer<byte[], byte[]> producer, String streamTaskId) {
...
+                        } else {
+                            String prefix = String.format("task [%s]", streamTaskId);
+                            log.error(String.format("%s Error sending record to topic %s", prefix, topic), exception);
+                        }
+                    }
+                });
+                return;
+            } catch (TimeoutException e) {
+                if (attempt == MAX_SEND_ATTEMPTS) {
+                    throw new StreamsException(String.format("failed to send record to topic %s after %d attempts", topic, attempt));
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollector.java]: Add the stream task id prefix here as well for both exception message and the warning log entry.

**Review by guozhangwang — COMMENTED** [review]
Minor comments otherwise LGTM.

**Comment by dguy** [comment]
@guozhangwang updated. Thanks

**Pull request closed without merging**